src/training.py

```diff
@@ -93,15 +157,30 @@ def training_step(self, batch, batch_idx):
 
     def validation_step(self, batch, batch_idx):
         loss, embeddings = self._forward_batch(batch, return_embeddings=True)
-        # save predictions for visualization
+        emb_np = embeddings.cpu().detach().numpy()
+        evt_ids_np = batch[1].cpu().numpy()
+        cluster_preds = self.clustering.cluster_and_link(emb_np, evt_ids_np)
+
+        metrics_results = {}
+
+        for metric in self.metrics:
+            if isinstance(metric, (F1ScoreMetric, PrecisionScoreMetric, RecallScoreMetric, AccuracyScoreMetric)):
+                metric.update(evt_ids_np, cluster_preds)
+            elif isinstance(metric, BaseScoreMetric):
+                metric.update(emb_np, cluster_preds)
+
+            metric_name = type(metric).__name__
+            self.epoch_metrics[metric_name].append(metric.compute())
```

Comment:
not sure that this should work in a such manner. Better to avoid this boilerplate code by using a proper realization of compute and update. E.g.
<img width="900" alt="image" src="https://github.com/t3hseus/spd-timeslice-unravelling/assets/12282410/8065c160-8d7c-44be-bcde-0f58832fd488">

Reply:
https://torchmetrics.readthedocs.io/en/stable/pages/lightning.html